Donut Chart directive, styles, ng-docs, and unit test

dtaylor113 opened the pull request:
The styles for the donut chart need to live in angular-patternfly#master (in new angular-patternfly.css file) since the donut chart implementation is being generated by angular-patternfly (there is no equivalent in patternfly#master)

waldenraines commented on dist/styles/angular-patternfly.css:
```diff
@@ -0,0 +1,29 @@
+.donut-chart-pf {
+	width: 100%;
```
This spacing seems a little extreme.  Can we use 2 or 4?

dtaylor113 replied:
Yeah, sorry, that was a cut & paste.  I'll change it to 2.

dtaylor113 replied:
fixed

dtaylor113 replied:
Hi Walden, 
Silly question.  I'm doing a code review and I click on the blue + to add a comment to a piece of code; how did you get the snippet to appear here in the Conversations, above the comment? 
Thanks,
- Dave

waldenraines commented on src/charts/c3/c3.mixin.js:
```diff
@@ -0,0 +1,57 @@
+angular.module('patternfly.charts').factory('ChartsMixin', function() {
```
The name of the file or the factory name should match.  This should probably be `charts-mixin.factory.js` to match the established pattern.

waldenraines replied:
Should this be a factory?  I feel like you could accomplish the same thing using a constant.  Where is the title set from outside of this factory?

dtaylor113 replied:
I believe it should be a factory.  The version we have in our cfmePOC actually has all the defaults for all the C3 chart types, so the file is injected into donut, sparkline, utilization, etc.. charts.
I will change it's filename.
-Thanks

waldenraines replied:
> The version we have in our cfmePOC...

You'll have to excuse my ignorance but what is cfmePOC?

You can inject an [angular constant](https://docs.angularjs.org/api/auto/service/$provide#constant) as well.  To me it makes sense to make changing the title dynamic but all of these static config values make me think of a constant.  I'll have to take a deeper look at how this is used as I don't see where the title is set at first glance.

dtaylor113 replied:
Sorry, the cfmePOC is a dashboard for containers where we initially developed a lot of the c3 charts (trends, utilization, donut, sparkline, etc.).  We are slowly moving these charts into a-pf#master.

dtaylor113 replied:
Hi, changed this to 'angular.module('patternfly.charts').constant('c3ChartDefaults', {...' and renamed file accordingly.'

waldenraines commented on src/charts/c3/c3.mixin.js:
```diff
@@ -0,0 +1,57 @@
+            };
+      };
+
+    var getDefaultDonutSize = function() {
```
It's best practice to declare all `var`s at the top of the function.

dtaylor113 replied:
Sorry, I'm not understanding this request.  Could you please provide an example?
Thanks

waldenraines replied:
Sorry, what I mean is this:

```javascript
angular.module('patternfly.charts').factory('ChartsMixin', function() {
  var getDefaultDonut, getDefaultDonutSize;

  getDefaultDonut = function () {...};
  getDefaultDonutSize = function () {...};

};
```

This is actually equivalent to what you have currently once it's parsed.  The trouble is that by not declaring your `var`s up top you can run into oddities like this

```javascript

var blah = 1;
function someFunction () {
	if (!blah) {
		var blah = 10;
	}
	console.log(blah);
}
someFunction();
```

This results in 10.  Now this is obviously a contrived example but it's a best practice to declare `var`s up top, outright to avoid any funny business (javascript is already funny enough).

dtaylor113 replied:
Ok, that is a good reason.  Thanks for explaining.

dtaylor113 replied:
Fixed

waldenraines commented on src/charts/c3/c3.mixin.js:
```diff
@@ -0,0 +1,57 @@
+    'use strict';
+
+    // Donut Chart Defaults
+    var getDefaultDonut = function(title) {
```
Don't we usually put a space after the `function` keyword ala `function (title) {`

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+ *
+ * @description
+ *   Directive for rendering a percentage used donut/radial chart.  The Used Percentage fill starts at 12 o’clock and
+ *   moves clockwise.  Whatever portion of the donut not Used, will be represented as Available, and rendered as a
```
Typo

>  Whatever portion of the donut not Used

Should probably be "Whatever portion of the donut is not used..."

dtaylor113 replied:
Fixed

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+ </file>
+ </example>
+ */
+angular.module('patternfly.charts').directive('pfDonutPctChart', ['ChartsMixin', '$timeout',
```
This is obviously outside the scope of this PR but we should look into using [grunt-ng-annotate](https://github.com/mzgol/grunt-ng-annotate) so we don't have to use this very verbose syntax ([`'dependency, function (dependency) { ...`.

See https://github.com/patternfly/angular-patternfly/issues/34

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+      scope: {
+        id: '@',
+        config: '=',
+        data: '=',
```
I think it's best practice to use prefixed attributes (i.e. `chart-id`, `chart-data`) so as to avoid a potential collision with actual HTML elements.  If you want to use the html ID of the element you could use the `$attrs` parameter that is passed into the controller.

dtaylor113 replied:
Hi, I added: 

controller: ['$scope', '$attrs',
        function($scope, $attrs) {
          var donutTooltip;

          $scope.donutChartId = 'donutChart';
          if ($attrs.id) {
            console.log('$attrs.id'+$attrs.id);
            $scope.donutChartId = $attrs.id + $scope.donutChartId;

...but am now getting an error in the unit tests:

PhantomJS 1.9.8 (Linux) Directive: pfDonutPctChart should trigger no threshold FAILED
	Error: [$injector:unpr] Unknown provider: $attrsProvider <- $attrs
	http://errors.angularjs.org/1.3.17/$injector/unpr?p0=%24attrsProvider%20%3C-%20%24attrs

I temporarily removed the unit test, and the actual code works fine, but once I restore the unit test, it fails.  I'm googling the issue, but not sure what is causing this.

waldenraines replied:
It's `controller: function($scope, $element, $attrs, $transclude, otherInjectables) { ... },`

https://docs.angularjs.org/api/ng/service/$compile

dtaylor113 replied:
I am going to move this back into a scope.config property which it was originally.  I was trying to reuse the 'id' attr, but it is causing too much issues.

dtaylor113 replied:
Fixed

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+            return color;
+          };
+
+          var donutTooltip = function(scope) {
```
It's best practice to declare the `var` at the top of the controller to avoid any potential hoisting issues.

dtaylor113 replied:
Hi, what var are you referring too, color & percentUsed?
Are you suggesting that it should be something like:
$scope.statusDonutColor = function(scope) {
            var color, percentUsed;
...
...
}

waldenraines replied:
In this case I'm talking about `var donutTooltip = function(scope) {`.  I personally like having those at the top and the `$scope` functions at the bottom but regardless the `var`s should be declared at the top (since they will be hoisted there anyway).

dtaylor113 replied:
Fixed

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+              } else {
+                return '#0088CE';
+              }
+            }
```
This logic is very confusing to me because of the negative condition and the multiple return points.  I suggest reworking it like this (untested):

```javascript
$scope.getStatusColor = function(used, thresholds) {
    var color = '#0088CE';

    if (thresholds) {
        if (used >= thresholds.error) {
            color = '#CC0000';
        } else if (used >= thresholds.warning) {
            color = '#EC7A08';
        }
    }
 
    return color;
};
```

dtaylor113 replied:
Fixed, thanks for the code snippet

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+            var color = {
+              pattern: []
+            };
+            var percentUsed = scope.data.used / scope.data.total * 100.0;
```
Same note about declaring all `var`s at once at the top of the function.

waldenraines replied:
```javascript
$scope.statusDonutColor = function(scope) {
   var color, percentUsed;

   color = ....
};
```

or 

```javascript
$scope.statusDonutColor = function(scope) {
  var percentUsed = scope.data.used / scope.data.total * 100.0,
      color = {
        pattern: []
      };
};

Although the latter looks a little strange here.
```

dtaylor113 replied:
Fixed

waldenraines commented:
Do we have a coding standard on `var blah = function(variable){};` vs `var blah = function (variable) {};`?  I see both in this PR.

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+                  return '<span class="donut-tooltip-pf" style="white-space: nowrap;">' +
+                            Math.round(d[0].ratio * 100) + '%' + ' ' + $scope.config.units + ' ' + d[0].name +
+                         '</span>';
+                }
```
I personally prefer this style, as I find it much more readable:

```javascript
var result = false;

if (condition) {
    result = true;
}

return result;
```

dtaylor113 replied:
Hi, you wrote: "Do we have a coding standard on var blah = function(variable){}; vs var blah = function (variable) {};? I see both in this PR."
Honestly, it took me a few re-reads to figure out the difference between A vs B (a single space!?).   Are we really going to be doing code reviews to this level of detail?  If so, code reviews are going to take much longer to do, and I'm not sure holding developers to such ridged standards is productive, or really makes the code any better.
I agree with many of the other best practice issues you have brought up, but I think we need to come to an agreement on what best practices it makes sense to follow.
Thanks,
- Dave

waldenraines replied:
I don't think we'll have to point these out in a code review once I turn on `white: true` in `.jshintrc` since the builds will fail instead :stuck_out_tongue_winking_eye:

waldenraines replied:
> I don't think we'll have to point these out in a code review once I turn on white: true in .jshintrc since the builds will fail instead

See https://github.com/patternfly/angular-patternfly/pull/36

dtaylor113 replied:
We've been wanting to speak to you about the .jshintrc for the ridged indentation as well ;-)  
It makes commenting out code for testing really difficult.  I guess I just never worked on a project where there was such close attention to formatting.  I can understand best practices for naming conventions, var declarations, etc., maybe even indentation, but spaces in method signatures!?   IMO this is not an efficient use of our time and it really doesn't make the code any more readable.

waldenraines replied:
> We've been wanting to speak to you about the .jshintrc for the ridged indentation as well ;-)

Haha, I'm certainly open to the discussion.  I can see the points on both sides.

> It makes commenting out code for testing really difficult.

Can you elaborate on this please?  

> I guess I just never worked on a project where there was such close attention to formatting. I can understand best practices for naming conventions, var declarations, etc., maybe even indentation, but spaces in method signatures!? IMO this is not an efficient use of our time and it really doesn't make the code any more readable.

Ultimately, I suppose that's where we disagree.  I believe having consistency does help make code more readable.

Although we may want to consider taking this discussion to the mailing list?

jeff-phillips-18 replied:
Can we allow these discrepancies for this PR and fix them if/when https://github.com/patternfly/angular-patternfly/pull/36/ gets accepted? If we decide to enforce these rules, then these issues can be fixed at that time.

dtaylor113 replied:
In regards to the indentation making debugging difficult.  I often want to comment out an outer function, such as a $timeout, in order to run the code immediately:

         // $timeout(function() {
            var donutChartTitle = element[0].querySelector('text.c3-chart-arcs-title');
            ...
         // }, 300);

I comment out the outer function, then 'grunt build', and of course the .jshintrc fails this, then have to go back, fix indentation, rerun, the do all over again when I restore the outer function.
I realize I can temporarily comment out the property in .jshintrc, but that is just another step to remember to restore before PR.

jeff-phillips-18 replied:
Dave, you can always use --force on the grunt commands which will continue execution regardless of jshint of test failures.

waldenraines replied:
> Can we allow these discrepancies for this PR and fix them if/when #36 gets accepted? If we decide to enforce these rules, then these issues can be fixed at that time.

Definitely.  I don't want to hold up this PR on something as of yet undecided.

I'll start a thread on the mailing list about this because it's outside of the context of this particular PR and when this is merged I have no problem running my regexes again to fix the spacing for #36.

> ... comment out the outer function, then 'grunt build'...

What if we used the grunt live reload server like most other grunt-based projects?  Is there a reason why we are building every single time we need to test a change?

P.S.: Check out [my website's gruntfile](https://github.com/waldenraines/waldenrainesdotcom/blob/master/Gruntfile.js#L28-L61) for an example (based off of yeoman's [generator-angular](https://github.com/yeoman/generator-angular)), particularly the watch config.  It runs  jshint on javascript files when they are changed but that doesn't kill the webserver.  It also reloads the web browser on changes and runs tests on test file changes.

dtaylor113 replied:
Hi, irt. 'What if we used the grunt live reload server like most other grunt-based projects? Is there a reason why we are building every single time we need to test a change?"  
When I'm working on a directive, I use the ng-docs 'dist/docs' to view changes, even style changes.   Don't we need to do a full build in order to get a fully updated 'dist/docs'?

dtaylor113 replied:
irt. "P.S.: Check out my website's gruntfile" 
"...watch config. It runs jshint on javascript files when they are changed but that doesn't kill the webserver."
-- awsome! that's what I want in angular-patternfly

"It also reloads the web browser on changes, "
-- think a-pf already does this

"runs tests on test file changes, "
-- this would be nice to have

"and automatically updates asset includes when they are added/removed from the bower.json file using grunt-wiredep."
-- this is nice, but probably a separate discussion.

So, should I change the watch config to not stop on jshint errors and also rerun karma tests upon change, create a PR.  Or do you usually run the intent by the team first before coding?

Thanks,
- Dave

waldenraines replied:
I'm actually working on updating the grunt file right now.  You're right that usemin and wiredep are pretty useless for this project since it's not a typical webapp (with an index.html, etc.).

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+          var donutTooltip = function(scope) {
+            return {
+              contents: function (d) {
+                if (scope.config.tooltipFn) {
```
What should happen here if `scope.config.tooltipFn` is falsy?  Is it an error condition?

dtaylor113 replied:
scope.config.tooltipFn is an optional property

waldenraines replied:
:+1:

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+            return {
+              columns: [
+                ['Used', scope.data.used],
+                ['Available', scope.data.available]
```
Again, the lack of translations is outside the scope of this PR but it made me think of this:  https://github.com/patternfly/angular-patternfly/issues/35

waldenraines commented on src/charts/donut/donut-pct-chart-directive.js:
```diff
@@ -0,0 +1,354 @@
+
+          $scope.config = $.extend(true, chartsMixin.getDefaultDonutConfig(), $scope.config);
+          $scope.updateAll($scope);
+        }],
```
Shouldn't the alignment of this be even with the `link` below?

dtaylor113 replied:
Link is in alignment with templateUrl and controller properties above, which also indent their contents by 2 spaces. LGTM?

waldenraines replied:
I'm talking about the `}]` above.  It's difficult to tell once it's been commented on but I would think the end of the controller block (`}]`) and the `link` property right below it (the controller and the link are in the same column) should also be in the same column.

[Look at the code](https://github.com/dtaylor113/angular-patternfly/blob/addDonutChart/src/charts/donut/donut-pct-chart-directive.js#L305-L306) before my comment, it's easier to see what I mean.

dtaylor113 replied:
lol, I tried to align it:

 }],
 link: function(scope, element, attrs) {

But am now getting:

Linting src/charts/donut/donut-pct-chart-directive.js ...ERROR
[L312:C7] W015: Expected '}' to have an indentation at 9 instead at 7.
      }],

waldenraines replied:
Hah.  It's because of the verbose style of dependency injection (i.e. the []).  It would probably be happy if you kept the } at column 9 and put the ] on the next line at column 7 (since that would match the function declaration above).

waldenraines commented on styles/angular-patternfly.css:
```diff
@@ -0,0 +1,29 @@
+.donut-chart-pf {
+	width: 100%;
```
Same thing here with the spacing.

waldenraines commented:
Missing tests for the mixin factory.

jeff-phillips-18 commented:
LGTM

dtaylor113 commented:
Hi Walden,
I believe I have addressed all of your issues that are within the scope of this PR.
Thanks for all the feedback,
- Dave
p.s. I will squash all commits into a single one before merge

waldenraines commented on dist/styles/angular-patternfly.css:
```diff
@@ -0,0 +1,29 @@
+  font-size: 12px;
+  font-weight: normal;
+  color: #333;
+}
```
Shouldn't this be minified?

dtaylor113 replied:
There is a minified version in 'dist/styles/angular-patternfly.min.js'

waldenraines replied:
Yeah, sorry, wrong file!

waldenraines commented:
LGTM, thanks!

erundle commented on styles/angular-patternfly.css:
```diff
@@ -0,0 +1,29 @@
+.donut-title-big-pf {
+  font-family: "Open Sans", Helvetica, Arial, sans-serif;
+  font-size: 22px;
+  font-weight: normal;
```
We need to use font-weight: 400 instead of normal. The way the fonts are currently declared in the font file are a bit weird and normal isn't mapped to anything.

What is the reason for needing the specify the font-family? That should be on the body and picked up by default.

erundle replied:
The rest looks good so after you have looked at this let me know and I can merge.

dtaylor113 replied:
Ok, fixed those css issues. thanks
